### ISSUE TITLE
Q: Best way to handle back-navigation parameters?

### ISSUE THREAD

**Issue by omnoms** [winjs/winjs#920]
I've now built a whole view/page that I would like to use like a controller.. basically when a user has selected items in a SeZo-calendar, I want to return those to the previous view when the back() navigation occurs from that page.

What's best practice here? I could just create a globally scoped namespace that holds the calendar values (which I am doing now), but is that really the only way? Can I capture when a user has returned to a view by navigating backwards?

I'm thinking in terms of separation of concerns that the calendar page shouldn't hold logic depending on which view it was navigated to but rather be dumb and just return a value in the back-navigation event...

**Comment by xirzec** [comment]
The intent of Pages was that all transient state would be passed in the 'state' object provided by the Navigation toolkit. So your controller could use the Navigation state as a model, update it as necessary, and reset the view when the user navigates back based on what's inside the state.

**Comment by omnoms** [comment]
Ok so I've tried this;

```javascript
        unload: function () {
             //TODO: Respond to navigations away from this page.
            var lv = document.querySelector(".cl_ZoomedIn");
            if (lv && lv.winControl) {
                var items = lv.winControl.selection.getItems();
                var tArr = [];
                items._value.forEach(function (item) {
                    tArr.push(item.data.locDate);
                })
                var tObj = {
                    "View": "Calendar",
                    "Data": tArr
                }
                WinJS.Navigation.state = tObj;
            }
        },
```

and then to try and catch it in the ready-statement of the hub, which is the current page it navigates back to when going back (for now), I've added this simple line;

```javascript
console.log(WinJS.Navigation.state);
```

and that shows that the state is undefined when navigating back. I've had log-lines in the unload section so I know it populates the object properly and sets the navigation.state... but it seems that navigating backwards resets the state... ?

**Comment by omnoms** [comment]
It seems that the WinJS.Navigation.state is not ready/accessible during the page's ready-statement... if I do a check on WinJS.Navigation.state once the page is fully rendered, then it is populated with the values I've passed to it in the unload-function.

What event would be best then to listen to in order to consume the navigation.state ?

**Comment by omnoms** [comment]
However.... setting WinJS.Navigation.state prior to calling WinJS.Navigation.back() DOES seem to reset the state...

```javascript
        calendarSingleClickHandler: WinJS.Utilities.markSupportedForProcessing(function(ev) {
            var nav = WinJS.Navigation;
            var lv = document.querySelector(".cl_ZoomedIn");
            if (lv && lv.winControl) {
                var item = lv.winControl.itemDataSource.list.getAt(ev.detail.itemIndex);
                calendarView.clicked = true;
                nav.state = { "View": "Calendar", "Date": item.locDate }
                console.log(nav.state);
                nav.back();
            }
        }),
```

**Comment by xirzec** [comment]
@omnoms Your page's `ready` function should get the state passed in as the second parameter - you shouldn't have to reach into WinJS.Navigation.state yourself. You might also have more luck adding your custom data onto the state object instead of replacing it entirely.

**Comment by omnoms** [comment]
The second parameter is still undefined when navigating backwards with WinJS.Navigation.back() despite setting or merging the state-object with my new properties/values.

I have a console.log() on the second parameter in the ready-function of the hub-page. When loading the app.. it's outputting undefined. As is my console.log(nav.state) in the hub.onloadingstatechanged function..

I navigate to the calendar and it's passing on a "From" property in the state-object that I can read in the calendar (this is a remnant, but also useful for troubleshooting). I merge new data into the state object and console.log that prior to running nav.back() and it shows the state object merged with what was there previously. When the hub is rendered the first undefiend is output from what I presume is the ready-statement and then later when all is complete, the last undefined is outputted from the hub.onloadingstatechanged function...

So basically the nav.back() is overwriting/resetting the state-object... why?

I have no issues if I click the back-button in the top-left corner but I want the single-click function to be a quick-way to set the current date-range to the week it was clicked in and navigate back..

**Comment by omnoms** [comment]
What I noticed if I do the back-navigation with the button first, so that it picks up the selection items on the first go is that it seems to cache the state object set from the unload the first time around.

The unload has a if-statement that says it won't set the state object if the single-click was performed, and I can see that on the second pass it's not setting the values, but despite that, the hub's readystatement and onloadingstatechanged is outputting the old state-values rather than the new ones.

**Comment by omnoms** [comment]
App started, navigated to HUB-page 
```
HUB.ReadyStatement Options: undefined
HUB.OnLoadingStateChanged: undefined
```
Navigate to the calendar page, right-click a couple of dates and click the back-button in the upper left-corner
```
HUB.ReadyStatement Options: undefined
HUB.OnLoadingStateChanged: {"View":"Calendar","Data":["‎2015‎-‎02‎-‎02","‎2015‎-‎02‎-‎03"]}
Multi-day selection:‎2015‎-‎02‎-‎02,‎2015‎-‎02‎-‎03
```
Navigate back to the calendar, single-click a date that runs afformentioned code and merges the stateobject with the previous state values. It outputs the state-object before running .back()
```
[object Object]
   {
      [functions]: ,
      __proto__: { },
      Date: "‎2015‎-‎02‎-‎03",
      From: "HUB",
      View: "Calendar"
   }
HUB.ReadyStatement Options: {"View":"Calendar","Data":["‎2015‎-‎02‎-‎02","‎2015‎-‎02‎-‎03"]}
HUB.OnLoadingStateChanged: {"View":"Calendar","Data":["‎2015‎-‎02‎-‎02","‎2015‎-‎02‎-‎03"]}
Multi-day selection:‎2015‎-‎02‎-‎02,‎2015‎-‎02‎-‎03
```

**Comment by omnoms** [comment]
I set a console.log in the unload statement of the calendar-page, and the ready-statement of the hub is kicked off before the unload is even running. So the ready-function of the hub could never hold the value at the first pass of the function.

Here's the chain of events from that log-output;

```
HUB.ReadyStatement Options: undefined
HUB.OnLoadingStateChanged: undefined

HUB.ReadyStatement Options: undefined
Setting nav.state in unload
HUB.OnLoadingStateChanged: {"View":"Calendar","Data":["‎2015‎-‎02‎-‎03","‎2015‎-‎02‎-‎04"]}
Multi-day selection:‎2015‎-‎02‎-‎03,‎2015‎-‎02‎-‎04

[object Object]
   {
      [functions]: ,
      __proto__: { },
      Date: "‎2015‎-‎02‎-‎03",
      From: "HUB",
      View: "Calendar"
   }

HUB.ReadyStatement Options: {"View":"Calendar","Data":["‎2015‎-‎02‎-‎03","‎2015‎-‎02‎-‎04"]}
HUB.OnLoadingStateChanged: {"View":"Calendar","Data":["‎2015‎-‎02‎-‎03","‎2015‎-‎02‎-‎04"]}
Multi-day selection:‎2015‎-‎02‎-‎03,‎2015‎-‎02‎-‎04
```

**Comment by omnoms** [comment]
Even more verbose, complete log.
```
HTML1300: Navigation occurred.
File: default.html
DOM7011: The code on this page disabled back and forward caching. For more information, see: http://go.microsoft.com/fwlink/?LinkID=291337
File: default.html
HUB.ReadyStatement Options: undefined
Deprecation warning: moment().subtract(period, number) is deprecated. Please use moment().subtract(number, period).
HUB.OnLoadingStateChanged: undefined
Calendar Ready Statement
HUB.ReadyStatement Options: undefined
Setting nav.state in unload
HUB.OnLoadingStateChanged: {"View":"Calendar","Data":["‎2015‎-‎02‎-‎03","‎2015‎-‎02‎-‎04"]}
Multi-day selection:‎2015‎-‎02‎-‎03,‎2015‎-‎02‎-‎04
Calendar Ready Statement
Calendar Single-Click
[object Object]
   {
      [functions]: ,
      __proto__: { },
      Date: "‎2015‎-‎02‎-‎03",
      From: "HUB",
      View: "Calendar"
   }

Calling nav.back()
HUB.ReadyStatement Options: {"View":"Calendar","Data":["‎2015‎-‎02‎-‎03","‎2015‎-‎02‎-‎04"]}
HUB.OnLoadingStateChanged: {"View":"Calendar","Data":["‎2015‎-‎02‎-‎03","‎2015‎-‎02‎-‎04"]}
Multi-day selection:‎2015‎-‎02‎-‎03,‎2015‎-‎02‎-‎04
```

**Comment by omnoms** [comment]
As a workaround I'm passing whatever data I want to pass back into a temporary object in the calendar view/page and just use the unload function to set the navigation state to whatever that is since the .back() function seems to reset the state-object entirely I basically populate the state object after .back() has done it's work.

**Comment by omnoms** [comment]
And lastly, since not all pages have a main controller like the HUB with an onloadingstatechanged, I've been forced to listen to the WinJS.Navigation event "navigated" to ensure that the state is readable since it's not in the page's ready-statement. 

So I've resorted to binding and unbinding a view/page specific function onto that event which in turn just copies the state-object's values to view-specific implementation of the calendar's returned values. This way I keep the logic of handling the calendar's returned values in the view it's pertinent to.

**Comment by xirzec** [comment]
Sorry @omnoms, I didn't understand your original request. So your problem is you want to pass some state to the previous page when navigating backwards. `WinJS.Navigation.state` is intended to store transient state of a page so that the page can rehydrate when a user hits back, that's why you are seeing `Navigation.back()` remove the state. It's actually resetting `state` back to the value that was set for `WinJS.Navigation.state` when that page was navigated away from.

In general if you want to pass new state to a Page you pass it along with your call to `WinJS.Navigation.navigate` but there isn't a built-in pattern for passing state backwards in the navigation history. To accomplish this, you'll have to use some global app mechanism for storing state.

**Comment by omnoms** [comment]
In the same spirit as global app namespace/variable/mechanism, I used the unload function of the calendar view that it's navigating backwards from. The Unload event/function is called *after* the navigation has occurred by .back() and the state is overwritten but before the navigated event is fired off.

So I'll live with how it's implemented now and use the navigated event to listen to the statechanges from the calendar. Perhaps later I'll fire off my own custom events from the calendar instead.
